pkgs/native_assets_cli/test/validator/validator_test.dart

```diff
@@ -0,0 +1,395 @@
+      package: config.packageName,
+      name: 'foo.dylib',
+      architecture: config.targetArchitecture,
+      os: config.targetOS,
+      linkMode: DynamicLoadingBundled(),
+    ));
+    final result = await validateBuild(config, output);
+    expect(result.success, isFalse);
+    expect(
+      result.errors,
+      contains(contains('has no file')),
+    );
+  });
+
+  for (final (linkModePreference, linkMode) in [
+    (LinkModePreference.static, DynamicLoadingBundled()),
+    (LinkModePreference.dynamic, StaticLinking()),
+  ]) {
+    test('native code asset wrong linking $linkModePreference', () async {
+      final config = BuildConfig.build(
+        outputDirectory: outDirUri,
+        packageName: packageName,
+        packageRoot: tempUri,
+        targetArchitecture: Architecture.arm64,
+        targetOS: OS.iOS,
+        targetIOSSdk: IOSSdk.iPhoneOS,
+        buildMode: BuildMode.release,
+        linkModePreference: linkModePreference,
+        supportedAssetTypes: [NativeCodeAsset.type],
+        linkingEnabled: false,
+      );
+      final output = BuildOutput();
+      final assetFile = File.fromUri(outDirUri.resolve('foo.dylib'));
+      await assetFile.writeAsBytes([1, 2, 3]);
+      output.addAsset(
+        NativeCodeAsset(
+          package: config.packageName,
+          name: 'foo.dart',
+          file: assetFile.uri,
+          linkMode: linkMode,
+          os: config.targetOS,
+          architecture: config.targetArchitecture,
+        ),
+      );
+      final result = await validateBuild(config, output);
+      expect(result.success, isFalse);
+      expect(
+        result.errors,
+        contains(contains(
+          'which is not allowed by by the config link mode preference',
+        )),
+      );
+    });
+  }
+
+  test('native code wrong architecture', () async {
+    final config = BuildConfig.build(
+      outputDirectory: outDirUri,
+      packageName: packageName,
+      packageRoot: tempUri,
+      targetArchitecture: Architecture.arm64,
+      targetOS: OS.iOS,
+      targetIOSSdk: IOSSdk.iPhoneOS,
+      buildMode: BuildMode.release,
+      linkModePreference: LinkModePreference.dynamic,
+      supportedAssetTypes: [NativeCodeAsset.type],
+      linkingEnabled: false,
+    );
+    final output = BuildOutput();
+    final assetFile = File.fromUri(outDirUri.resolve('foo.dylib'));
+    await assetFile.writeAsBytes([1, 2, 3]);
+    output.addAsset(
+      NativeCodeAsset(
+        package: config.packageName,
+        name: 'foo.dart',
+        file: assetFile.uri,
+        linkMode: DynamicLoadingBundled(),
+        os: config.targetOS,
+        architecture: Architecture.x64,
+      ),
+    );
+    final result = await validateBuild(config, output);
+    expect(result.success, isFalse);
+    expect(
+      result.errors,
+      contains(contains(
+        'which is not the target architecture',
+      )),
+    );
+  });
+
+  test('native code no architecture', () async {
```

Comment:
Why do we allow `NativeCodeAsset` not to specify an architecture to begin with?

Reply:
In dry run, we only have the OS as input, no architecture.